### DISCUSSION TITLE
Metatopic: Next Steps, Strategic Plan

### DISCUSSION THREAD

**Discussion by ampli** [opencog/link-grammar#1134] [Ideas]
From https://github.com/opencog/link-grammar/issues/785#issuecomment-766431171:
> How do we evaluate the quality of the learning? This last question is key. The answer is to (1) create a random dictionary (2) create a random corpus of text from the random dictionary (3) apply the learning algo to learn a dictionary (4) compare the learned dictionary to the starting dict of (1): if its a good match, then learning worked. If not, then not.

Can the current English dict take the role of such a dict, when sentences that have a full parse serve as such "random text"?
Such sentences can be easily "generated" in any desired quantity by filtering existing text (like stories and documents).

> The goal of learning is to deduce a dictionary using probabilistic techniques. What are the best learning algos? The best techniques?

Is there an algo that exploits the nesting feature of languages, i.e., that many words can be replaced by phases?

**Comment by linas** [comment]
> Can the current English dict take the role

No, not really. Proper testing of the algorithm requires testing to see how it does as the following are varied:
* Size of vocabulary
* Number of link types
* Fraction of words having 1,2,3.. connectors
* Fraction of words having ambiguous meanings (to see, to cut (saw))
* Size of training corpus

The idea is to start small, so that experiments can be run quickly.

>  Is there an algo

This is brand-new research. There aren't any algos, besides those you can dream up.

**Comment by ampli** [comment]
> Fraction of words having ambiguous meanings (to see, to cut (saw))

What leads to the hope that one can, in general, distinguish between such words by pure analysis of mutual info of word pairs in text?
After all, one can see a table and also saw (cut) a table. It seems that in general, a deeper "understanding" of text is needed for that.
("Deeper" is maybe mutual information of many words at once.)

>- Size of vocabulary
>- Number of link types
>- Fraction of words having 1,2,3.. connectors
>- Fraction of words having ambiguous meanings (to see, to cut (saw))
>- Size of training corpus

Can the mutual info algo actually reconstruct very small dicts?
Say 10 words, 5 links, words can have 1 or 2 links?

**Comment by ampli** [comment]
I edited my previous comment because it was not clear.

**Comment by linas** [comment]
> What leads to the hope that one can, in general, distinguish between such words by

There's a PDF on this

> by pure analysis of mutual info of word pairs in text?

That's not the correct "algorithm"

>  is maybe mutual information of many words at once

For example, the mutual information between a word and a disjunct.  Step 1) create word pairs Step 2) MST parse Step 3) extract disjuncts Step 4) compute word-disjunct pairs Step 5) classify word-disjunct pairs Step 6) create LG dict Step 7) evaluate accuracy of LG dict That's the current algo I use. Atnton screwed around with replacing step 2 and step 5 with neural nets.

> Can the algo actually reconstruct very small dicts?

Anton Kolonin and crew did this very quickly, with 100% accuracy, almost immediately. Where they got stuck was with "child-directed speech" because they were unable to control the variables. Thus, they were unable to get meaningful measurements; they couldn't interpret their results.  (Well, they gave up too easy; they were hoping to find magic-potion/philosophers-stone in a month or two, and gave up after discovering baking-soda+vinegar.)

> the mutual info algo

There's no such "algo". That's like saying "the cosine distance algo" Everywhere you use cosine distance, you can use mutual info. Or you can use any other distance metric of your choosing.  The only advantage for MI is that there are assorted formal proofs that it is the "best" for probabilities, and also there are many decades of experience with maximum entropy principles. Which one is free to ignore, but then one needs some other conceptual foundations for what one is doing.  Axel Kleidon has some inspirational writings.. well, I guess so do many others.

**Comment by linas** [comment]
Ooops, the PDF word-sense disambiguation is at .... I'm having trouble finding it. Chapter 6, page 39 of https://github.com/opencog/learn/blob/master/learn-lang-diary/connector-sets-revised.pdf mentions this briefly; somewhere there is another PDF that expands this into greater detail. 

Its not particularly complicated.  Its part of classification. All cutting verbs have similar disjuncts and can be classified that way. All sensing verbs (hearing, touching) have similar disjuncts, and are classified together. Something like "saw" will have a pile of disjuncts that are cutting-like, and other disjuncts that are sensing-like, and the code tries to factorize this into those two classes, assigning a different "meaning" to each. It resembles sparse-matrix factorization; I can explain more when you're ready.  All that the PDF said was this, plus some worked examples. 

The decision to split a collection of disjuncts into two meanings or not is just a standard classification decision. I was using agglomerate clustering. So was Anton. He kept using K-means, I kept telling him not to, because there is no way to know what K should be. Before he figured this out, he went full-on neural net, and got worse results. ... but one cannot be sloppy about such things. Boeing 747's aren't large paper airplanes made out of metal.  That's not how things scale.

**Comment by ampli** [comment]
It is indeed clearly possible if you know the set of disjuncts that are observed of each word,  to find out that some words have more than one meaning. However, I don't have, for now, any idea how to improve/extend this algo or how to use it better.

But there are other things I don't understand.

>2\) MST parse Step
>...
>Atnton screwed around with replacing step 2 and step 5 with neural nets.

How you MST-parse English not using any kind of ML?

And how do you make MST parsing on sentences you have generated according to your own dict?

**Reply by linas** [reply]
MST is short for "maximum/minimum spanning tree". It can be obtained for any kind of graph, whenever you have weighted edges. One chooses edges so that the sum of the weights is minimized/maximized. (There's a wikipedia article on this; first done for electric power line planning in Hungary(?) in the 1930's(?), the weight was the geographic distance.)

In the present case, the edges are pairs of words, so if you can associate weights to word-pairs, you can create an MST parse. To get a planar parse, you can enforce an additional no-links-crossing rule.  What should the weights be? There are generic arguments that the entropy of the word pair is "the best choice" (the entropy is the "mutual information") One reason its a good choice is that entropies are naturally additive. Another reason is that assorted theorems state that maximizing entropy is the best way to extract information. This is "trivial" only in textbook examples, but promptly gets non-trivial in real-world examples. There are books on "maximum entropy principles".  Applying MST to natural language parsing is described in Deniz Yuret PhD thesis. The thesis is neither very long, nor is it very complicated; I recommend that you read it, I think you will not have many problems understanding it. (There are one or two locations where he makes a few "magic leaps", which bugged me, but I eventually found the explanations. And then forgotten them...)  ... the first step is to obtain MI for word-pairs.

> how do you make MST parsing on sentences you have generated according to your own dict?

Step 0: generate a random dict
Step 1: use the dict to generate N sentences
Step 2: use those sentences to compute the MI of word pairs This is done by extracting word-pairs that lie within a window of each other, typically W=6. One counts the word pairs, After counting, the MI is easily computed. See Yuret for details
Step 3: Use the MI to perform MST parses
Step 4: extract disjuncts from the MST parses
Step 5: cluster together similar disjuncts, to obtain word-classes, and link types.
Step 6: compare the results of step 5 to the original grammar.

The "science questions" are "how does accuracy depend on N?" or rather, "what is the optimum value for N given a vocabulary of size V"? And more subtle questions: what if the grammar has many disjuncts with only one connector, and only a few with 2 connectors? What if the vocabulary has many disjuncts with 2 or 3 connectors, and only a few with 1 connector? What happens if the grammar contains many synonyms? Or only a few? All of these parameters affect learning speed and accuracy.  

The hardest step in the above is step 5, because there are many ways of doing clustering.  I can explain the few that I explored.

Oh, and one more: "step 7: automate steps 0-6"  In some ways, this is the hardest step. Running the above pipeline is like watching a laundry machine do your laundry. If you don't watch, you come back later to discover you used too much soap, and there's bubbles all over he floor. Or you used hot water when you should have used cold. Or the machine just broke, got unplugged, tipped over ... it's mind-bogglingly boring and repetitive and difficult to automate. Especially when you find and fix bugs in earlier stages: you have to decide whether to restart, or whether to keep going with imperfect data. This is the number 1 reason I stopped working on this: its difficult and unrewarding.

FYI, I have code for steps 2-5. Its not fast or efficient, but it is flexible; it is easy for me to change things around. I started work on step 1, but its .. surprisingly complicated. The prototype works, for tiny dicts.

**Reply by linas** [reply]
MI is defined as follows. Let `x`,`y` be two words, and let `N(x,y)` be the number of times that the pair `(x,y)` was observed in real life. Observations are done by scanning text, and simply counting, whenever a pair of words occur near each other (where "near to" means "up to some max distance").

Then  
```
MI(x,y) = log_2 [ P(x,y) / (P(*,y) P(x,*)) ]
```
where the probability (frequency) is
```
P(x,y) = N(x,y) / N(*,*)
```
and `*` is "wild-card".  Thus `N(*,*)` is the total number of observed word-pairs and `N(x,*)` is the total number of word pairs, with word `x` on the left. That is, `N(x,*) = sum_y N(x,y)`. Therefore,
```
P(x,*) = N(x,*) / N(*,*) = sum_y P(x,y)
```
and so on.

**Reply by linas** [reply]
... and the whole point of Yuret's thesis was to define the above, to describe MST parsing, to explain why this is a good idea, give some examples, and then run some real-world tests proving that the idea actually works.

You want to get a PhD? Do steps 0-7 and write up the results. We can get your favorite nearby university to perform the formal ceremony and grant the degree. Accomplishing the above is bona-fide PhD work; no professor will argue that it's less than that. Getting it published would be a real achievement. I'll work with whomever to make this happen.

**Reply by ampli** [reply]
> I started work on step 1, but its .. surprisingly complicated.

Before I say something (later) about the rest of things, I would like to mention that I think I know how to do that using the current LG library code with a few modifications. Using the current English dict, I guess it may use much memory and CPU for constructing a "parse set" for sentences of N words when N is not very small.  From such a "parse set", random sentences can be generated.

If this is interesting I can try to provide a demo.

**Reply by linas** [reply]
> f this is interesting I can try to provide a demo.

Yes!  But (first) I would like to understand how this works ... what is a "parse set"? Why do you need to specify the length of the sentence? 

For a test dict, you can use one of the other dicts -- the german dict is medium-sized, the lithuanian dict is small but is "different" in structure. LG version 1.0 might have a small english dict -- I thought there was a small one, somewhere at some point.

I implemented something called "odometer enumerated aggregation" but I don't want to contaminate your brain with that. (I'll explain it later) I suspect that there are other, wildly different strategies for generating sentences, but I can't think of any, right now. Given that I never really understood the "parse set", maybe you've got a good one ... lets figure that out .. explain that to me.

**Reply by ampli** [reply]
>I would like to understand how this works ... 

It is based on the idea described in my post "[Fill in the gaps](https://groups.google.com/g/link-grammar/c/Q9i5rkVbuPY/m/e4xRQfT3BAAJ)".

If you use several such wildcard words you get all the sentences for these number of words.
In the "old times" this was extremely slow.  Now I guess it would be faster, but still very slow. However, several vast optimizations are possible: 1. When reading the dict, in order to reduce the number of needed expressions.  2. When discarding duplicate disjunct, in order to collapse together disjuncts of different words.  3. When building disjuncts (new algo). 4. When extracting the links (i.e. actually generating the sentences). 5. May need to use lowercase letters after ID for idioms to reduce the number of uppercase links. 6. Maybe defining idiom links as length=1 would help too.

The "parse set" is just the data structure used in `extract-links.c`.

>Why do you need to specify the length of the sentence?

It could actually be "all sentences up to length N" if I would mark all words as optional. But this would add very much overhead. I don't have any other idea how I can do it with really variable number of words.

**Reply by linas** [reply]
Oh, OK, that seems reasonable. At this time, a tool that works slowly is better than no tool at all. My tools take days-weeks wall-clock time to run, so a few days or longer to generate a corpus is fine.  So, sure, write a demo!

**Reply by ampli** [reply]
>Step 2: use those sentences to compute the MI of word pairs
Step 3: Use the MI to perform MST parses

Are these steps considered as "solved"? Do you have a program that does that?

**Reply by ampli** [reply]
>Are these steps considered as "solved"? Do you have a program that does that?

>>I have code for steps 2-5.

So the answer is "yes".

Other questions:
1. What are the sentence length s that are to be used?
2. Are MST parses for English correspond well enough to manual parsing, or is it not important?

**Reply by linas** [reply]
> I have code for steps 2-5.

It's in https://github.com/opencog/learn The README's will explain how to install and run it. The final output is an LG SQLite file. However, there's a lot of howevers; I'd have to explain those. Stage 5 is the one that needs tuning and exploration. Yes, re-writing stages 2-4 in pure C/C++ would probably speed them up by a factor of 10x or more. However, I tried to create something generic, something that could work for biology (for example), and not just LG. 

> What are the sentence length s that are to be used?

Arbitrary; not constrained. Don't care.

> Are MST parses for English correspond well enough to manual parsing

Yes. That is Yuret's thesis: he shows they correspond "well enough". However, he (and everyone else) gave up at that point, and never pushed it any further. That is because they didn't know about the concept of disjuncts. It never occurred to them that mining for disjuncts is the next step. I strongly believe that mining for disjuncts will strongly improve quality, and it will surpass the hand-created dictionaries. But this is a claim/belief, the whole point of the project is to prove that this is true.

**Comment by ampli** [comment]
> write a demo!

My first step was to port the old modification to the latest LG, without any optimization.

A test on the `de` dict with a sentence length of 12 words (excluding `LEFT_WORD`/`RIGHT_WORD`) with `-limit=10000000` (10 million) finished creating the parse-set structure in 194 seconds and needed additional 3780 seconds to extract all the sentences and write them to a file (5.6GB!). In addition to the bottlenecks that I already mentioned, there was a bottleneck in `link_lists()` when it needed to repetitively scan `Parse_choice` linked lists of tens of thousands of elements (this, of course, can be fixed, and will benefit also the regular use of the library).

With the `en` `4.0.dict` I had less success. With more than 3 words (excluding `LEFT_WORD`/`RIGHT_WORD`) the process is OOM-killed when it reaches ~13GB of memory during the disjunct build. However, this can be fixed by order of magnitudes in several ways even without changing the disjunct building algo (which inherently has a low memory consumption).

The generated sentences include regex  `NAME` and `CAPITLIZED_WORD`, often more than once. I guess it may be a good idea to suppress their generation.
Another observation is that often `LEFT_WALL` and `RIGHT_WALL` appear in the middle of sentences. I guess they should be suppressed there.
I hoped to get `.` in the middle of sentences. This could be an easy way to generate shorter sentences (than the given fixed number of words). Apparently, the `de` dict doesn't include this possibility (it can be easily fixed for that). The algo can also have a preference for including `.` (and even only one `.`) in sentences in order to produce sentences of different lengths in certain amounts.
Another possibility is to add API to use certain given words.

I will now try a few of my suggested optimizations to see if I can generate long-enough sentences from the `en` dict in a reasonable time.

**Reply by linas** [reply]
Don't bother with `en` (except maybe as a stress-test of a giant dict)  FWIW, I have a machine with 256GB; in practice, up to 200 GB is usable, the rest is needed for "other stuff".   

I really want t start with small dictionaries (a thousand words, 10-20 link types ... or maybe even much smaller for the initial debug/bringup.) and see if those can be learned (with high accuracy). If that works, the next steps are larger dictionaries, and more complex dictionaries. I except both CPU and RAM bottlenecks, and all sorts of difficulties scaling.

**Reply by linas** [reply]
I started writing code to generate random dicts. I might be done in a day or two!

**Reply by linas** [reply]
Done.. I can now create very basic dictionaries. Code is in https://github.com:opencog/learn/tree/master/fake

Here's an example:
```
a: <a>;
<a>: (J-) or (J+ & B+);

b c d e f g: <b>;
<b>: (F-) or (A- & A+) or (G+ & A-);

h i j k: <c>;
<c>: (B+ & A+ & B+) or (A+) or (D- & A+ & C+);

l m n o p q: <d>;
<d>: (B-);

r s t u v w: <e>;
<e>: (A+ & J-) or (A+ & A-) or (A- & F+) or (D+) or (A+ & E- & B+) or (A+) or (D+ & A- & H+) or (E- & A+) or (A+ & A+ & A+) or (B- & A+ & I+) or (A- & C-) or (C+ & B-) or (A-) or (A+ & A-) or (G+) or (A+ & G-) or (B-) or (D-) or (A+ & E- & A+);

x y z aa ab: <f>;
<f>: (I+);

ac: <g>;
<g>: (C+ & B-);

ad: <h>;
<h>: (D- & B+ & A+) or (J+) or (J-) or (G-) or (A+);

ae: <i>;
<i>: (A-) or (F+) or (C+) or (E+) or (E- & A- & D-) or (E+) or (H-) or (A- & B- & B-);

af ag ah ai aj ak al: <j>;
<j>: (H- & B- & D-) or (H-);
```

A big issue, right now, is that it is not clear to me if every disjunct can get used. The disjuncts are randomly generated, and its not clear if there exist any sentences at all that use some of them. This will be an issue for learning: if a disjunct cannot be used while generating a corpus, it obviously can't be learned, and so the learned dict must not be marked as being "inaccurate" because of this.

I wouldn't be surprised if most disjuncts above are unusable.

To speed up corpus generation, it might be best to have only one word per word-class. Corpus post-processing can then randomly insert synonyms.

**Reply by linas** [reply]
So I thought about this a bit more, and it became clear that, to generate a corpus, one only needs a single representative word per word-class. So, in the above example, `af ag ah ai aj ak al` are all synonyms for a single "representative" word, call it `rep-j`. To generate a corpus, just generate a valid sentence with `rep-j` in it. In a second stage, one can trivially create 7 different sentences just by string-replacing `rep-j` with each of `af ag ah ai aj ak al`.

Multiple word meanings can be handled similarly. If the dictionary contains
```
<pos-j>: (B+) or (B+ & A- & B+) or (E+) or (C+ & A- & B+);
repj: <pos-j>;
<pos-e>: (A+) or (C+ & A- & F+);
repe: <pose>;

<wcl-x>: <pos-j> or <pos-e>;
ax ay az: <wcl-x>;
```
then it is enough to generate sentences with `repj` and `repe` in them, and then do simple string replacements to paste `ax ay az` in for each `repj` or `repe`.  That is,  `ax ay az` are synonyms, and they each have two word-senses or word-meanings: they have the "j" sense and the "e" sense. 

I've updated my code to generate dictionaries similar to this.

**Reply by ampli** [reply]
(Writing again after I did `page back` by mistake and the input-box text didn't get restored on return, apparently because the "view more" mechanism that clashes with restoring it.)

>To speed up corpus generation, it might be best to have only one word per word-class. Corpus post-processing can then randomly insert synonyms.

To this idea I referred above as:
>1. When reading the dict, in order to reduce the number of needed expressions.

I called it word category (according to the terminology in the library doc). The Brandom insertion of the words is done in `linkage_create()`.

>Multiple word meanings can be handled similarly. 

I implemented that in a more general way, by this method:
 
>2. When discarding duplicate disjunct, in order to collapse together disjuncts of different words.

(This WIP is almost complete.)

>A big issue, right now, is that it is not clear to me if every disjunct can get used.

In my implementation, the disjuncts that cannot be used can be found by observing which disjuncts are missing in the parse-set (the data structure that is created by `mk_parse_set()`.

**Reply by linas** [reply]
> word category 

I'm having trouble finding good names for the concepts. I keep trying to invent new names cause none of them fit well. But yes, I've called them "word classes" elsewhere. 

> I implemented that 

OK. I can certainly generate the random dictionaries in all sorts of different formats. If some particular one is easier, I can do that.

>  This WIP is almost complete

Cool let me know. I'm eager to try it (even as I know this will me a years-long project.) Just feeling excited, for the moment.

**Reply by ampli** [reply]
It is referred to as "category" [here](https://www.abisource.com/projects/link-grammar/dict/introduction.html#:~:text=to%20list%20all%20the%20words%20in%20a%20particular%20category,expression) and in the rest of the document (search for "category").

**Comment by linas** [comment]
See also discussion in #1290
